### PR TITLE
feat(stages/webhook) Webhook stage

### PR DESCRIPTION
**Pull request by amanya**
This PR is the implementation of the UI of the webhook stage discussed in https://github.com/spinnaker/spinnaker/issues/1512

This is a screenshot of the execution details view:

image

And the stage config:

image

@spinnaker/reviewers PTAL

### CODE REVIEW

File: app/scripts/modules/core/help/helpContents.js [hunk 1]
```diff
@@ -527,4 +527,14 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
     'markdown.examples': 'Some examples of markdown syntax: <br/> *<em>emphasis</em>* <br/> **<b>strong</b>** <br/> [link text](http://url-goes-here)',
     'application.enableRestartRunningExecutions': 'When this option is enabled, users will be able to restart pipeline stages while a pipeline is still running. ' +
     'This behavior can have varying unexpected results and is <b>not recommended</b> to enable.',
+    'pipeline.config.webhook.payload': 'Json payload to be added to the webhook call.',
+    'pipeline.config.webhook.waitForCompletion': 'If not checked, we consider the stage succeeded if the webhook return an HTTP status code 2xx, otherwise it will be failed. If checked, it will poll a status url (defined below) to grasp the progress of the stage.',
```
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: Please s/return/returns

Also, 'grasp the progress of the stage' sounds off to me - how about 'determine the progress of the stage'?

File: app/scripts/modules/core/help/helpContents.js [hunk 2]
```diff
@@ -527,4 +527,14 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
     'markdown.examples': 'Some examples of markdown syntax: <br/> *<em>emphasis</em>* <br/> **<b>strong</b>** <br/> [link text](http://url-goes-here)',
     'application.enableRestartRunningExecutions': 'When this option is enabled, users will be able to restart pipeline stages while a pipeline is still running. ' +
     'This behavior can have varying unexpected results and is <b>not recommended</b> to enable.',
+    'pipeline.config.webhook.payload': 'Json payload to be added to the webhook call.',
```
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: Please change Json to JSON and JsonPath to JSONPath here and below.

File: app/scripts/modules/core/help/helpContents.js [hunk 3]
```diff
@@ -527,4 +527,14 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
...
+    'pipeline.config.webhook.waitForCompletion': 'If not checked, we consider the stage succeeded if the webhook return an HTTP status code 2xx, otherwise it will be failed. If checked, it will poll a status url (defined below) to grasp the progress of the stage.',
+    'pipeline.config.webhook.useStatusUrlFromLocationHeaderTrue': 'Pick the status url from the Location header of the webhook\'s response call.',
+    'pipeline.config.webhook.useStatusUrlFromLocationHeaderFalse': 'Pick the status url from a json returned by the webhook\'s response call.',
```
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: s/a Json/the JSON

File: app/scripts/modules/core/help/helpContents.js [hunk 4]
```diff
@@ -527,4 +527,14 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
...
+    'pipeline.config.webhook.statusUrlJsonPath': 'JsonPath where to find the status url in the webhook\'s response json. (i.e. <samp>$.buildInfo.url</samp>)',
```
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: These might read easier as 'JSON path to the status URL'

File: app/scripts/modules/core/help/helpContents.js [hunk 5]
```diff
@@ -527,4 +527,14 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
...
+    'pipeline.config.webhook.statusJsonPath': 'JsonPath where to find the status information in the webhook\'s response json. (i.e. <samp>$.buildInfo.status</samp>)',
+    'pipeline.config.webhook.progressJsonPath': 'JsonPath where to find the progress information in the webhook\'s response json. (i.e. <samp>$.buildInfo.progress</samp>)',
```
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: What does progress mean (vs. status)?

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.html [hunk 6]
```diff
@@ -0,0 +1,97 @@
+<div class="form-horizontal">
+  <stage-config-field label="Your service's URL">
```
Comment (danielpeach) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.html]: I don't know if we have any fields that are prefixed with "Your". Maybe just "Payload URL" would be better here?

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.html [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  </stage-config-field>
+  <stage-config-field label="Method">
+    <ui-select ng-model="$ctrl.stage.method" class="form-control input-sm">
+      <ui-select-match placeholder="Select a method...">{{$select.selected}}</ui-select-match>
+      <ui-select-choices repeat="method in $ctrl.methods | filter: $select.search">
+        <span ng-bind-html="method | highlight: $select.search"></span>
+      </ui-select-choices>
+    </ui-select>
+  </stage-config-field>
+  <stage-config-field label="Payload" help-key="pipeline.config.webhook.payload">
```
Comment (danielpeach) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.html]: This field should probably be hidden for a GET request.

File: version.json [hunk 8]
```diff
@@ -1,4 +1,4 @@
 {
     "version": "n/a",
-    "created": 1461949989729
+    "created": 1489429408125
```
Comment (danielpeach) [inline version.json]: Please omit these changes.

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      this.command.errorMessage = e.message;
+    }
+  }
+
+  public waitForCompletionChanged(): void {
+    this.stage.waitForCompletion = this.viewState.waitForCompletion;
+  }
+
+  public useStatusUrlFromLocationHeader(): void {
+    this.stage.useStatusUrlFromLocationHeader = this.viewState.useStatusUrlFromLocationHeader;
```
Comment (danielpeach) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts]: Is there any reason for having a separate property on `viewState` that just gets mapped to a value on the stage?

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  public command: ICommand;
+  public viewState: IViewState;
+  public methods: string[];
+
+  constructor(public stage: any,
+              $scope: IScope,
+              private jsonUtilityService: JsonUtilityService) {
+    this.methods = ['GET', 'POST', 'PUT'];
```
Comment (danielpeach) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts]: These can be set directly on `methods` above (`public methods = ['GET', 'POST', 'PUT']`). Also..why not other HTTP methods?

File: app/scripts/modules/core/help/helpContents.js (hunk 5 above)
Comment (amanya) [inline app/scripts/modules/core/help/helpContents.js]: Progress mean a descriptive error message to be shown to the user, I rewrote it as: `JSON path to a descriptive message about the progress in the webhook\'s response JSON. (e.g. <samp>$.buildInfo.progress</samp>)` WDYT?

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts (hunk 10 above)
Comment (amanya) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookStage.ts]: I added additional most common methods, but if I define them above they are not shown in the ui...

**Comment by tomaslin** [comment]
from a ease of use standpoint, would be nice if there was a `try it out` button that will show the results of the webhook call next to it so that the json path values could be tested. Similar to to how expresssions evaluate to results.

**Comment by amanya** [comment]
@tomaslin that's a good idea, do you think it should be implemented just in frontend, or better do it in orca to get the expression language resolved and such. Doing in orca also will need to create an endpoint in gate, right?

**Comment by tomaslin** [comment]
you won't be able to resolve any expressions in the configuration phase of the pipeline. In the ideal world it would use the same code so you wouldn't have to replicate it.

**Comment by amanya** [comment]
@tomaslin yeah, as it is a somewhat big feature I think it would be best to leave it for a second round, if you agree I will add a list of TODO features in the issue (https://github.com/spinnaker/spinnaker/issues/1512) with your comment. WDYT?

**Comment by tomaslin** [comment]
@amanya yeah, I'm happy for that to be a feature enhancement. Great job on this btw

File: app/scripts/modules/core/help/helpContents.js (hunk 5 above)
Comment (danielpeach) [inline app/scripts/modules/core/help/helpContents.js]: Sounds good.

File: app/scripts/modules/core/help/helpContents.js [hunk 11]
```diff
@@ -527,4 +527,15 @@ module.exports = angular.module('spinnaker.core.help.contents', [])
...
     'This behavior can have varying unexpected results and is <b>not recommended</b> to enable.',
+    'pipeline.config.webhook.payload': 'JSON payload to be added to the webhook call.',
+    'pipeline.config.webhook.waitForCompletion': 'If not checked, we consider the stage succeeded if the webhook returns an HTTP status code 2xx, otherwise it will be failed. If checked, it will poll a status url (defined below) to determine the progress of the stage.',
+    'pipeline.config.webhook.statusUrlResolutionIsGetMethod': 'Use the webhook\'s URL with GET method as status endpoint.',
+    'pipeline.config.webhook.statusUrlResolutionIsLocationHeader': 'Pick the status url from the Location header of the webhook\'s response call.',
+    'pipeline.config.webhook.statusUrlResolutionIsWebhookResponse': 'Pick the status url from the JSON returned by the webhook\'s response call.',
+    'pipeline.config.webhook.statusUrlJsonPath': 'JSON path to the status url in the webhook\'s response JSON. (i.e. <samp>$.buildInfo.url</samp>)',
+    'pipeline.config.webhook.statusJsonPath': 'JSON path to the status information in the webhook\'s response JSON. (i.e. <samp>$.buildInfo.status</samp>)',
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/help/helpContents.js]: The `i.e.`s should probably be `e.g.`s

File: app/scripts/modules/core/pipeline/config/stages/webhook/webhookExecutionDetails.html [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    <div class="well alert-info" ng-if="ctrl.stage.context.progressMessage || ctrl.stage.status">
+      <h4>Results</h4>
+      <dl class="dl-narrow dl-horizontal ng-scope">
+        <dt>Status</dt>
+        <dd class="ng-binding">{{ctrl.stage.status}}</dd>
+        <dt>Info</dt>
+        <dd class="ng-binding">{{ctrl.stage.context.progressMessage}}</dd>
+      </dl>
+      <p>
+      </p>
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/pipeline/config/stages/webhook/webhookExecutionDetails.html]: Any need for this empty `<p>` tag?

**Comment by anotherchrisberry** [comment]
A couple of super minor cosmetic things, LGTM otherwise

**Comment by amanya** [comment]
@anotherchrisberry fixed!

**Comment by anotherchrisberry** [comment]
Just needs a rebase and we'll get this merged

**Comment by amanya** [comment]
@anotherchrisberry done

**Comment by anotherchrisberry** [comment]
Awesome, thanks again for putting this together!

**Pull request merged**